src/RadixDlt.NetworkGateway.PostgresIntegration/Services/QueryHelper.cs

```diff
@@ -0,0 +1,110 @@
+ * or use made or its functionality or any performance or operation of any code or protocol
+ * caused by bugs or programming or logic errors or otherwise);
+ *
+ * A. any offer, purchase, holding, use, sale, exchange or transmission of any
+ * cryptographic keys, tokens or assets created, exchanged, stored or arising from any
+ * interaction with the Work;
+ *
+ * B. any failure in a transmission or loss of any token or assets keys or other digital
+ * artefacts due to errors in transmission;
+ *
+ * C. bugs, hacks, logic errors or faults in the Work or any communication;
+ *
+ * D. system software or apparatus including but not limited to losses caused by errors
+ * in holding or transmitting tokens by any third-party;
+ *
+ * E. breaches or failure of security including hacker attacks, loss or disclosure of
+ * password, loss of private key, unauthorised use or misuse of such passwords or keys;
+ *
+ * F. any losses including loss of anticipated savings or other benefits resulting from
+ * use of the Work or any changes to the Work (however implemented).
+ *
+ * You are solely responsible for; testing, validating and evaluation of all operation
+ * logic, functionality, security and appropriateness of using the Work for any commercial
+ * or non-commercial purpose and for any reproduction or redistribution by You of the
+ * Work. You assume all risks associated with Your use of the Work and the exercise of
+ * permissions under this License.
+ */
+
+using Microsoft.EntityFrameworkCore;
+using RadixDlt.NetworkGateway.Abstractions;
+using RadixDlt.NetworkGateway.GatewayApi.Exceptions;
+using RadixDlt.NetworkGateway.PostgresIntegration.Models;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace RadixDlt.NetworkGateway.PostgresIntegration.Services;
+
+internal static class QueryHelper
+{
+    internal static async Task<Dictionary<EntityAddress, long>> ResolveEntityIds(ReadOnlyDbContext dbContext, List<EntityAddress> addresses, GatewayApiSdk.Model.LedgerState ledgerState, CancellationToken token)
+    {
+        var entities = await dbContext
+            .Entities
+            .Where(e => e.FromStateVersion <= ledgerState.StateVersion && addresses.Contains(e.Address))
+            .AnnotateMetricName()
+            .ToDictionaryAsync(e => e.Address, e => e.Id, token);
+
+        return entities;
+    }
+
+    internal static async Task<TEntity> GetEntity<TEntity>(ReadOnlyDbContext dbContext, EntityAddress address, GatewayApiSdk.Model.LedgerState ledgerState, CancellationToken token)
+        where TEntity : Entity
+    {
+        var entity = await dbContext
+            .Entities
+            .Where(e => e.FromStateVersion <= ledgerState.StateVersion)
+            .AnnotateMetricName()
+            .FirstOrDefaultAsync(e => e.Address == address, token);
+
+        if (entity == null)
```

Comment:
Just remember that we can't use it anywhere virtual accounts/identities are expected! Are we OK with that?

Reply:
yes, we don't expect virtual entities in places where I've used it.